gridftp/gsi-wuftpd/source/support/man/ftw.3
.\"
.\" Copyright (c) 1999 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id$
.\"
.TH TREEWALK 3 August 4, 1988
.UC 4
.SH NAME
ftw, treewalk \- tree walk
.SH SYNOPSIS
.nf
.B #include <ftw.h>
.PP
.B ftw(path, function, maxfds)
.B char *path;
.B int (*function)(), maxfds;

.B treewalk(path, function, maxfds, options)
.B char *path;
.B int (*function)(), maxfds, options;
.fi
.SH DESCRIPTION
.PP
The \fItreewalk\fP call descends the directory hierarchy rooted in
\fIpath\fP, visiting each node in turn.  Visiting a node consists of
calling the routine \fIfunction\fP with three arguments.  The first
is the current path name, the second a pointer to a stat structure
containing information about the path name, and the third a flag
value.  The declaration of the function may be as follows:
.nf

	#include <sys/stat.h>

	function(path, st, flag)
	char *path;
	struct stat *st;
	int flag;
.fi
.PP
The \fIflag\fP value will be one the following values, as specified in
<\fIftw.h\fP>.
.PP
.RS
 FTW_F		a regular file
 FTW_D		a directory
 FTW_D2	second visit to a directory
 FTW_DNR	an unreadable directory
 FTW_NS	an unknown type, the stat failed
.RE
.PP
If the flag value is \fIFTW_NS\fP, the contents of the stat buffer are
undefined.
.PP
\fIMaxfds\fP governs the number of file descriptors that \fItreewalk\fP may
use.  If passed as -1, no limit is set.  Unless \fIfunction\fP needs
to open files, -1 is the suggested behavior.
.PP
\fIOptions\fP is a bit mask which changes the behavior of \fItreewalk\fP.
Its value is set by or'ing together any of the following values specified
in <\fIftw.h\fP>.
.PP
.RS
 FTW_CHDIR		\fItreewalk\fP may change directories
 FTW_DIRLAST		visit the node after its descendents
 FTW_SYMLINK	follow symbolic links
.RE
.PP
\fIFTW_CHDIR\fP allows \fItreewalk\fP to descend the tree using the
\fIchdir\fP(2) system call.  This makes \fItreewalk\fP run much faster.
A side effect of this flag is that all path names passed to \fIfunction\fP
are relative, not absolute.
.PP
By default, \fItreewalk\fP does a depth-first search, visiting each node
before any of its descendents.  If \fIfunction\fP returns any other value
than 0 or -1, \fItreewalk\fP will not continue, but will return this value
to its caller.  If the returned value is -1, \fItreewalk\fP will continue,
but no descendents of the current node will be visited.  The
\fIFTW_DIRLAST\fP flag causes \fItreewalk\fP to visit the node again after
all of its descendents have been visited.  The second visit to the directory
will call \fIfunction\fP with the \fIFTW_D2\fP flag instead of \fIFTW_D\fP.
.PP
If \fIFTW_SYMLINK\fP is set, \fItreewalk\fP will pass the result of a
\fIstat\fP(2) call to \fIfunction\fP, rather than an \fIlstat\fP(2) call.
In the case of symbolic links pointing to directories, \fItreewalk\fP
will continue down the directory tree pointed to by the link.  If
\fItreewalk\fP is unable to indirect through the link, the node will
be ignored and \fIfunction\fP will not be called.  It should also be
noted that if the symbolic links form a loop, \fItreewalk\fP will not
detect this condition.
.PP
If \fItreewalk\fP detects any error other than EACCES (see \fIFTW_NS\fP,
\fIFTW_DNR\fP, and \fIFTW_CHDIR\fP, above) or EMFILE (in the case of
running out of file descriptors) it returns -1 to its caller, leaving
\fIerrno\fP unchanged.
.SH BUGS
The \fIpath\fP string passed to \fIfunction\fP is the buffer currently
being processed by \fItreewalk\fP and must therefore be copied before
any modifications are made.
.SH HISTORY
The \fItreewalk\fP routine is a superset of the \fIftw\fP routine as
provided by System V.  \fIFtw\fP is available for compatibility.
.SH SEE ALSO
find(1), getdtablesize(2), stat(2), symlink(2), glob(3)


ChangeLog
2000-10-10  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_log.c: Removed my_vsnprintf() as it is now in
	string_funcs.c

	* verror.c: Removed my_vsnprintf() as it is now in
	string_funcs.c
	
	* string_funcs.c, string_funcs.h: Created these files.

	* myproxy_destroy.c (main): Bug fix - use created proxy file for
	authentication and not default proxy.
	(main): Bug fix - passphrase was not being correctly read.

2000-10-09  Von Welch  <vwelch@ncsa.uiuc.edu>

	* verror.c (my_vsnprintf): Fixed inverted logic in checking of
	HAVE_VSNPRINTF
	
2000-10-04  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_server.c (main): Print strerror() when failure occurs
	opening configuration file.

	* myproxy.c(parse_command): Fixed command_value type.

2000-10-03  Von Welch  <vwelch@ncsa.uiuc.edu>

	* ssl_utils.c: Fixes for openssl build with PEM callback
	differences (openssl has a user data parameter while ssleay does
	not).

	* ChangeLog: Created after 0.2alpha1 release



gridftp/gsi-wuftpd/source/support/ftw.c
/****************************************************************************  
 
  Copyright (c) 1999 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 by Stan Barber.
  Portions Copyright (c) 1997 by Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.org/license.html.
 
  $Id$
 
****************************************************************************/
#include "../src/config.h"

#ifdef HAVE_DIRENT_H
#include <dirent.h>
#else
#include <sys/dir.h>
#endif

#include <errno.h>
#include <string.h>
#include <stdlib.h>

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/stat.h>

#include "ftw.h"

#define	NODESC	-1

#ifdef HAVE_LSTAT
#define	ISLINK(sb)	((sb.st_mode&S_IFMT) == S_IFLNK)
#else
#define lstat stat
#endif

#define	ISDIR(sb)	((sb.st_mode&S_IFMT) == S_IFDIR)
#define	ISDOT(dp) \
	(dp->d_name[0] == '.' && (!dp->d_name[1] || \
	    (dp->d_name[1] == '.' && !dp->d_name[2])))

extern int errno;
static int g_fds, (*g_fn) (), g_opts;
static char *bp;

int treewalk(char *path, int (*fn) ( /* ??? */ ), int maxfds, int opts);

/*
 * cycle through the directories at the top of the tree, otherwise, once
 * you run out of descriptors you have to keep reusing the same one and
 * it gets *real* slow.
 */
typedef struct d_fd {
    struct d_fd *next;
    DIR *dirp;
    off_t off;
} FD;

static FD *freep, *node;

static int walk(register char *name)
{
#ifdef HAVE_DIRENT_H
    register struct dirent *dp;
#else
    register struct direct *dp;
#endif
    register int rval;
    struct stat sb;
    FD cur;
    char *save;

    if (!freep)
	freep = &cur;
    else
	node->next = &cur;
    node = &cur;
    cur.off = 0;

  getfd:if (!g_fds) {
	freep->off = telldir(freep->dirp);
	closedir(freep->dirp);
	freep = freep->next;
	++g_fds;
    }
    if (!(cur.dirp = opendir(bp))) {
	if (errno == EMFILE) {
	    g_fds = 0;
	    goto getfd;
	}
	return (errno == EACCES ? (*g_fn) (bp, &sb, FTW_DNR) : -1);
    }
    else
	--g_fds;

    for (; *name; ++name);
    *name++ = '/';
    for (rval = 0, dp = readdir(cur.dirp); dp; dp = readdir(cur.dirp)) {
	if (ISDOT(dp))
	    continue;
	(void) strcpy(name, dp->d_name);
	if (lstat(bp, &sb)) {
	    rval = errno == EACCES ?
		(*g_fn) (bp, &sb, FTW_NS) : -1;
	    if (rval)
		break;
	}
#ifdef HAVE_LSTAT
	if (ISLINK(sb) && g_opts & FTW_SYMLINK)
	    if (stat(bp, &sb))
		continue;
#endif
	if (!ISDIR(sb)) {
	    rval = (*g_fn) (bp, &sb, FTW_F);
	    if (rval)
		break;
	    continue;
	}
	if (g_opts & FTW_DIRLAST)
#ifdef HAVE_D_NAMLEN
	    save = name + dp->d_namlen;
#else
	    save = name + strlen(dp->d_name);
#endif
	rval = (*g_fn) (bp, &sb, FTW_D);
	if ((rval && rval != NODESC) || (rval = walk(name)))
	    break;
	if (g_opts & FTW_DIRLAST) {
	    *save = '\0';
	    rval = (*g_fn) (dp->d_name, &sb, FTW_D2);
	    if (rval)
		if (rval == NODESC)
		    rval = 0;
		else
		    break;
	}
	if (cur.off) {
	    *name = '\0';
	    if ((cur.dirp = opendir(bp))) {
		seekdir(cur.dirp, cur.off);
		/* tricky; if we have to reset the directory pointer we know
		 * it's the next one to reuse */
		freep = &cur;
		--g_fds;
	    }
	    /* directory moved from under us!!! */
	    else {
		rval = -1;
		break;
	    }
	}
    }
    closedir(cur.dirp);
    ++g_fds;
    return (rval);
}

static int chwalk(register char *name)
{
#ifdef HAVE_DIRENT_H
    register struct dirent *dp;
#else
    register struct direct *dp;
#endif

    register int rval;
    struct stat sb;
    FD cur;
    char *pwd, *getwd(char *);

    if (!freep)
	freep = &cur;
    else
	node->next = &cur;
    node = &cur;
    cur.off = 0;

    if (chdir(name))
	return (errno == EACCES ? (*g_fn) (name, &sb, FTW_DNR) : -1);

  getfd:if (!g_fds) {
	freep->off = telldir(freep->dirp);
	closedir(freep->dirp);
	freep = freep->next;
	++g_fds;
    }
    if (!(cur.dirp = opendir("."))) {
	if (errno == EMFILE) {
	    g_fds = 0;
	    goto getfd;
	}
	return (errno == EACCES ? (*g_fn) (".", &sb, FTW_DNR) : -1);
    }
    else
	--g_fds;

    for (rval = 0, dp = readdir(cur.dirp); dp; dp = readdir(cur.dirp)) {
	if (ISDOT(dp))
	    continue;
	if (lstat(dp->d_name, &sb)) {
	    rval = errno == EACCES ?
		(*g_fn) (dp->d_name, &sb, FTW_NS) : -1;
	    if (rval)
		break;
	}
	pwd = NULL;
#ifdef HAVE_LSTAT
	if (ISLINK(sb) && g_opts & FTW_SYMLINK) {
	    if (stat(dp->d_name, &sb))
		continue;
	    if (ISDIR(sb)) {
		/* NOSTRICT */
		if (!(pwd = malloc((u_int) MAXPATHLEN))) {
		    rval = -1;
		    break;
		}
		if (!getwd(pwd)) {
		    rval = -1;
		    break;
		}
	    }
	}
#endif
	if (!ISDIR(sb)) {
	    rval = (*g_fn) (dp->d_name, &sb, FTW_F);
	    if (rval)
		break;
	    continue;
	}
	rval = (*g_fn) (dp->d_name, &sb, FTW_D);
	if ((rval && rval != NODESC) || (rval = chwalk(dp->d_name)))
	    break;
	if (g_opts & FTW_DIRLAST) {
	    rval = (*g_fn) (dp->d_name, &sb, FTW_D2);
	    if (rval)
		if (rval == NODESC)
		    rval = 0;
		else
		    break;
	}
	if (pwd && chdir(pwd)) {
	    rval = -1;
	    break;
	}
	if (cur.off) {
	    if ((cur.dirp = opendir("."))) {
		seekdir(cur.dirp, cur.off);
		/* tricky; if we have to reset the directory pointer we know
		 * it's the next one to reuse */
		freep = &cur;
		--g_fds;
	    }
	    /* directory moved from under us!!! */
	    else {
		rval = -1;
		break;
	    }
	}
    }
    closedir(cur.dirp);
    ++g_fds;
    if (chdir(".."))
	return (-1);
    return (rval);
}

#ifndef HAVE_FTW
/* S5 compatible ftw(BA_LIB) */
int ftw(char *path, int (*fn) ( /* ??? */ ), int maxfds)
{
    return (treewalk(path, fn, maxfds, 0));
}
#endif

int treewalk(char *path, int (*fn) ( /* ??? */ ), int maxfds, int opts)
{
    struct stat sb;
    int rval;
    char *pwd, *getwd(char *);

    if (lstat(path, &sb))
	return (errno == EACCES ? (*fn) (path, &sb, FTW_NS) : -1);

    pwd = NULL;
#ifdef HAVE_LSTAT
    if (ISLINK(sb) && opts & FTW_SYMLINK) {
	if (stat(path, &sb))
	    return (0);
	if (ISDIR(sb)) {
	    /* NOSTRICT */
	    if (!(pwd = malloc((u_int) MAXPATHLEN)))
		return (-1);
	    if (!getwd(pwd))
		return (-1);
	}
    }
#endif
    if (!ISDIR(sb))
	return ((*fn) (path, &sb, FTW_F));

    if (!maxfds)
	return (-1);
    g_fds = maxfds == -1 ? getdtablesize() : maxfds;
    g_fn = fn;
    g_opts = opts;

    if (!(opts & FTW_CHDIR) && !(bp = malloc((u_int) MAXPATHLEN))) {
	errno = ENOMEM;
	return (-1);
    }
    rval = (*fn) (path, &sb, FTW_D);
    if (rval == NODESC)
	rval = 0;
    else if (!rval) {
	if (opts & FTW_CHDIR)
	    rval = chwalk(path);
	else
	    rval = walk(strcpy(bp, path));
	if (!rval && opts & FTW_DIRLAST) {
	    rval = (*fn) (path, &sb, FTW_D2);
	    if (rval == NODESC)
		rval = 0;
	}
    }
    if (pwd && chdir(pwd))
	return (-1);
    return (rval);
}
